### PR TITLE
IndexedField column name removal

### PR DESCRIPTION
**Pull request by sschmittBt**
Reworked IndexedField to remove the fixed column name and instead fetch the column name on query execution from the mapper

### CODE REVIEW

**Review by remi128 — CHANGES_REQUESTED** [review]
FieldContition line 135 has compiler error

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java [hunk 1]
```diff
@@ -32,13 +35,6 @@ public static IIndexedField create(String fieldName) {
...
   String getFieldName();
 
-  /**
-   * Return the name of the column in the datastore, usually the same as the field name
-   * 
-   * @return the column name
-   */
-  String getColumnName();
-
   static IIndexedField getIndexedField(String name, Class<?> pojoClass)
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java:50]: documentation a bit please

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java [hunk 2]
```diff
@@ -51,4 +47,22 @@ static IIndexedField getIndexedField(String name, Class<?> pojoClass)
           + "' must be static, final, and of type " + IIndexedField.class.getName());
   }
 
+  default String getColumnName(IDataStore<?, ?> datastore) {
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java]: documentation a bit please

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java [hunk 3]
```diff
@@ -51,4 +47,22 @@ static IIndexedField getIndexedField(String name, Class<?> pojoClass)
...
+    return getColumnName(datastore.getMapperFactory().getMapper(getClass()));
+  }
+
+  default String getColumnName(IMapper<?> mapper) {
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java]: documentation a bit please

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java [hunk 4]
```diff
@@ -51,4 +47,22 @@ static IIndexedField getIndexedField(String name, Class<?> pojoClass)
...
+    String fieldName = getFieldName();
+    String subFieldName = "";
+    int i = fieldName.indexOf('.');
+    if (i > 0) {
+      subFieldName = fieldName.substring(i);
+      fieldName = fieldName.substring(0, i);
+    }
+    IProperty field = mapper.getField(fieldName);
+    if (field == null)
+      throw new de.braintags.vertx.jomnigate.exception.NoSuchFieldException(getFieldName());
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java:92]: We have both exception types in the class: java.land.NoSuchFieldException and de.braintags.vertx.jomnigate.exception.NoSuchFieldException. Is that wished?

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java [hunk 5]
```diff
@@ -93,7 +92,7 @@ private Object handleSingleValue(Object value) {
     } else if (value instanceof Enum) {
       transformedValue = ((Enum<?>) value).name();
     } else if (value instanceof GeoSearchArgument) {
-      transformedValue = new JsonObject(Json.encode(value));
+      transformedValue = JsonObject.mapFrom(value);
     } else {
       try {
         // can not use datastore object mapper here because only the JSON datastore has the object mapper
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java:98]: isn't it useful to overwrite this method in MongoQueryExpression where you have access to the object mapper?

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/IndexedField.java [hunk 6]
```diff
@@ -11,34 +11,43 @@
...
+    return result;
   }
 
   @Override
-  public String toString() {
-    return fieldName;
+  public boolean equals(Object obj) {
+    if (this == obj)
+      return true;
+    if (obj == null)
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/IndexedField.java:53]: could it make sense to get one line like obj == null ||getClass() != obj.getClass ?

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/MappedIdFieldImpl.java [hunk 7]
```diff
@@ -12,40 +13,52 @@
...
+  public int hashCode() {
+    final int prime = 31;
+    int result = super.hashCode();
+    result = prime * result + ((mappedField == null) ? 0 : mappedField.hashCode());
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj)
```
Comment (remi128) [inline vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/MappedIdFieldImpl.java]: duplicate, cause following super call does the same

**Review by remi128 — CHANGES_REQUESTED** [review]

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java (hunk 4 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java:92]: I think it makes sense, the java one is thrown if the IndexedField is missing in the java class, and the jomnigate one is thrown when a field could not be found inside the mapper.

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java (hunk 5 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java:98]: MySql could also have a GeoSearchArgument, so we have to have a solution for both cases.
But nevertheless, this whole block will be removed in an upcoming change.

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/IndexedField.java (hunk 6 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/IndexedField.java:53]: I just use the auto generated equals method from Eclipse, but I don't think the compiler really cares in what way this expression is written.

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/MappedIdFieldImpl.java (hunk 7 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/MappedIdFieldImpl.java]: Fixed

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java (hunk 1 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java:50]: Added

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java (hunk 2 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java]: Added

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java (hunk 3 above)
Comment (sschmittBt) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/IIndexedField.java]: Added

File: vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java (hunk 5 above)
Comment (remi128) [reply vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java:98]: the mysql search expression for Geo differs a lot from the one of mongo, but ok, we will change that later together with the refactoring of mysql

**Review by remi128 — APPROVED** [review]

**Comment by remi128** [comment]
after resolviong conflicts merging ok

**Review by mpluecker — DISMISSED** [review]
Code look ok for me.

**Comment by mpluecker** [comment]
By the way, do we still need the IdField and IMappedIdField and their implementations? For example an IndexedField is not equals to a MappedIdFieldImpl even if they reporesent the same field of a mapped class.

**Comment by sschmittBt** [comment]
@mpluecker:
We have two use cases: 
Make a query to an ID field of a mapper:
`ISearchCondition.isEqual(mapper.getIdField(),"1234")`
Get some field properties of the ID field of a mapper, e.g. the property accessor:
mapper.getIdField().getPropertyAccessor().writeData(...)

We could make to seperate properties, getQueryIdField and getMappedIdField for example, but with this we can simply use getIdField().getField() to get to the mapped field.

**Comment by remi128** [comment]
ok, i did not really recognize intensively, that the api of IMapper was changed that way.

The api previously returned an IProperty for getIdField.
IProperty contains all information like fieldName, columnName etc. and has the information isIdField.
That way it contains all information already, which are now redundant inside the IMappedField and IIndexed field. Additionally the IMappedIdField contains the IProperty. 
Additionally the IMapper returns an IProperty for getField(), but an IMappedIdField - which is not an instance of IProperty - for getIdField(), so that we are mixing responsibilites of IMapper now, which are IMO not very logic.
Please find another solution here

**Comment by sschmittBt** [comment]
Continued in #95

**Pull request merged**